File: dart/leet_code/678.dart
/// Dart implementation of the LeetCode problem 678: Valid Parenthesis String.
///
/// This module contains a class `Solution` with a method `checkValidString`.
/// The `checkValidString` method takes in one parameter:
/// - `s`: A string representing the input string of parentheses.
///
/// The method returns a boolean indicating whether the input string is valid.
/// The string is valid if it can be made balanced by replacing '*' with '(' or ')' or an empty string.
///
/// Error Handling:
/// - This method assumes that the input parameter is well-formed, i.e., `s` is a string of parentheses and '*'.
/// - If the input parameter is not well-formed, the behavior of the method is undefined.
library;

import 'dart:math';

class Solution {
  /// Checks whether the input string of parentheses is valid.
  ///
  /// This method uses two counters to keep track of the minimum and maximum number of open parentheses.
  /// It iterates over each character in the input string. If the character is '(', it increments both counters.
  /// If the character is ')', it decrements both counters. If the character is '*', it decrements the minimum counter and increments the maximum counter.
  /// If the maximum counter is less than 0, it returns false. If the minimum counter is less than 0, it resets it to 0.
  /// After the iteration, if the minimum counter is 0, it returns true. Otherwise, it returns false.
  ///
  /// @param s The input string of parentheses.
  /// @return A boolean indicating whether the input string is valid.
  bool checkValidString(String s) {
    int low = 0, high = 0;
    for (int i = 0; i < s.length; i++) {
      if (s[i] == '(') {
        low++;
        high++;
      } else if (s[i] == ')') {
        low--;
        high--;
      } else {
        low--;
        high++;
      }
      if (high < 0) {
        return false;
      }
      low = max(low, 0);
    }
    return low == 0;
  }
}

File: go/leet_code/678.go
// Package leet_code provides solutions for LeetCode problems.
//
// This file provides a solution for the problem of checking if a string is valid.
// The string is valid if it can be made valid by inserting parentheses at any position.
// The function assumes that the string only contains the characters '(', ')', and '*'.
package leet_code

// checkValidString checks if a string is valid.
//
// It accepts one parameter:
// - s: a string that only contains the characters '(', ')', and '*'.
//
// The function returns a boolean indicating whether the string is valid.
//
// The function uses two counters to simulate the possible range of the number of open parentheses.
// It iterates over the string, updating the counters.
// If at any point during the iteration the right counter is less than zero, the function returns false.
// After the iteration, the function returns whether the left counter is zero.
//
// Time complexity analysis:
// - Best-case: O(n), when the string is valid.
// - Worst-case: O(n), when the string is not valid.
// - Average-case: O(n), as we always have to iterate over each character in the string.
func checkValidString(s string) bool {
	left, right := 0, 0
	for _, c := range s {
		if c == '(' {
			left++
		} else {
			left--
		}
		if c != ')' {
			right++
		} else {
			right--
		}
		if right < 0 {
			break
		}
		left = max(left, 0)
	}
	return left == 0
}

// max returns the maximum of two integers.
//
// It accepts two parameters:
// - a, b: the two integers to compare.
//
// The function returns the maximum of a and b.
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
